packages/selected-network-controller/src/SelectedNetworkController.ts

```diff
@@ -192,21 +185,34 @@ export class SelectedNetworkController extends BaseController<
 
     this.messagingSystem.subscribe(
       'NetworkController:stateChange',
-      (availableNetworkClientIds) => {
-        // if a network is updated or removed, update the networkClientId for all domains
-        // that were using it to the selected network client id
-        const { selectedNetworkClientId } = this.messagingSystem.call(
-          'NetworkController:getState',
-        );
-        Object.entries(this.state.domains).forEach(
-          ([domain, networkClientIdForDomain]) => {
-            if (!availableNetworkClientIds.includes(networkClientIdForDomain)) {
-              this.setNetworkClientIdForDomain(domain, selectedNetworkClientId);
-            }
-          },
+      ({ networkConfigurationsByChainId }, patches) => {
+        const patch = patches.find(
+          ({ op, path }) =>
+            (op === 'replace' || op === 'remove') &&
+            path[0] === 'networkConfigurationsByChainId',
         );
+
+        if (patch) {
+          selectedChainId = this.#getSelectedChainId();
+          Object.entries(this.state.domains).forEach(
+            ([domain, chainIdForDomain]) => {
+              if (
+                patch.op === 'remove' &&
+                !networkConfigurationsByChainId[chainIdForDomain]
+              ) {
+                // If the network was removed, fall back to the globally selected network
+                this.setChainIdForDomain(domain, selectedChainId);
+              } else if (
+                patch.op === 'replace' &&
+                chainIdForDomain === patch.path[1]
+              ) {
+                // If the network was updated, redirect to the network's default endpoint
+                this.setChainIdForDomain(domain, chainIdForDomain);
```

Comment:
We don't have a good way to inspect what rpc endpoint the network client proxy for the domain is currently pointed at, so we repoint the network client proxy for the domain at the default network client for the chainId any time the network configuration for that chainId changes